### DISCUSSION TITLE
This looks interesting

### DISCUSSION THREAD

**Discussion by ElixirHunter** [Cornerback24/Blender-NLA-Midi-Copier#5] [General]
Hello Cornerback,

I am very interested in the possibilities of this as I have been thinking of producing a Blender animation with objects reacting to MIDI messages.

I am a bit rubbish at computers and know nothing of python scripts but I did manage to install your work as an add on and made a box jump up and down when a certain note was played in the midi file.

I then made the box into an array around a curve to see if I could do the 'follow path' version (see picture below). I was hoping this would lead to each of the 88 boxes being allocated a MIDI note each and jumping up when that note played.

However, this did not work at all and despite printing out your manual and reading as much as is available I cannot get any further.

I am sure you are very busy but is there an 'idiots' guide to how I could do this?

Many thanks
Elixir

![image](https://user-images.githubusercontent.com/96600644/147300685-532e9ccf-452e-4474-8064-4e7f9dee660c.png)

**Comment by Cornerback24** [comment]
Hey, the way the Copy along path works is it animates separate objects, so once your boxes are separate objects, you should then be able to select all of them, choose your path in the Quick Copy Tools panel, and press Copy Action to Notes (the button in the Quick Copy Tools panel). 

If your boxes are not individual objects it won't work. Not sure exactly how you made the box array, but if for example you used an array modifier and curve modifier, you could do something like:

- Apply the array modifier
- Apply the curve modifier
- Select the boxes and go to edit mode
- Separate by loose parts (hotkey is P)
- Go back to object mode
- With all the boxes selected, set origin to geometry (can be done with the right click menu). (~~Might work without this but I imagine you'd want each box to have its own local origin~~. Actually I think this is necessary for the addon's path calculations.)

Then you'd have individual boxes, and you could use the Copy along path tool:
-  Select all the boxes
-  Choose your path in the Quick Copy Tools Panel
-  Click "Copy Action to Notes" in the Quick Copy Tools Panel

and I think that should work!

There's a simple example of just the copy along path part here https://youtu.be/6aVWFdhfHXg?t=373, it is a bit outdated (it uses an older version of this addon). Let me know if you still have issues with it.

**Comment by ElixirHunter** [comment]
Wow, thanks for the prompt and thorough response!

That works.

I am going to have a play around when I get time but I'm pretty sure I'll
be back to you with a few more questions at some point :-)

In the meantime, I hope you have a good Christmas.

Elixir

**Comment by ElixirHunter** [comment]
I hope you had a good Christmas Cornerback.

I have just spent some time making a MIDI test in Blender to see if this will be a workable add-on for my workflow.

Just chromatic scales to test every note of 88 on the piano and a reduced note for brass chords to test length of playing as well.
I emphasize - not proper music - Just a test!

You can see it here: [https://drive.google.com/file/d/1WX6ljBbkqnHDt0fm74SbDksaceH1nYnm/view?usp=sharing]


I am very interested to see if there is a way of changing 'both ends of the DATA stream'.

For example, making the velocity (volume of the note) and other elements of the MIDI data control objects.

.........and whether MIDI can control other aspects such as materials, lighting, for example rather than just actions.

I am sure I have missed a simple trick but could you have the position of a single object guided by the note. For example a box be at position x=200 when note C4  is played, 220 when C#4 is played 240 when D4 is played etc...

Many thanks in advance for any help/advice you can give.

Elixir

**Comment by Cornerback24** [comment]
That's a neat test you've got there.

There is some keyframe generation over in the Graph Editor that can generate different keyframes for different pitches. It also works with velocity and note length. It generates keyframes onto the selected property in the graph editor, and can be used on any property once the property has a keyframe on it (so that it can be selected in the graph editor).
Here's an example of using it to change an object's position based on pitch: https://drive.google.com/file/d/1U7GHHedsIoSHPiCe5QhgCavAA6p8IpEF/view?usp=sharing . 

The midi tools in the NLA Editor only copy actions, but actions can be used to animate things like lights as well. The Type needs to be changed to see those actions (for example an action that changes a light's power will be selectable once the Type is changed to Light).

**Comment by ElixirHunter** [comment]
Thanks.

I have tried some more experiments to see if I could control other elements as you suggest via the graph editor.

You can see the update here: [https://drive.google.com/file/d/1tg8JlbRU-EC65BSoTzhSXC1RKQUZCpHy/view?usp=sharing](url)

One thing I would really like to achieve but cannot see a way is to use a particle system which 'sprays out particles' only when there are notes played. Additionally, with the size, colour, emission strength of the particle being affected by volume or pitch.

I did try and do something in this update. The particle system is linked to the chromatic scale and changes colour as the scale rises (you can see it resetting back to blue at the beginning of each scale). However, I cannot find a way to turn it on and off easily.


Also on the brass chords - I deliberately recorded them with a massive diminuendo (decreasing velocity) and then crescendo (increasing velocity) thinking I might be able to link this velocity DATA to the 'amount' or 'influence' of an action. In short, when the brass are louder the cones expand more.

As usual, any ideas very welcome.

Elixir

**Comment by Cornerback24** [comment]
I'm not very familiar with the particle system in Blender, and so far I haven't found an easy way to turn off the particles either. If there's a solution that involves keyframing it could be done with using graph editor or maybe with scaled actions, but I'm not sure what that solution would be. 
Same for animating other aspects of the particles like size. You can add keyframes to particle properties, but it seems that also affects the particles once they are emitted. I'm not sure how to get around that either.
As far as copying actions to particle settings, it looks like the NLA editor midi tool is missing particle as a type, so I'll look into adding that. 

The addon doesn't currently have a way of controlling the action's influence, but that sounds like something I can look into adding. Maybe the addon could have an option to set the Animated Influence on action strips when the actions are copied.

**Comment by ElixirHunter** [comment]
Hello Again,

Sorry about the delay, been a bit busy.

I experimented with a few things and produced another animation. This is still for testing rather than musicality.
[https://drive.google.com/file/d/16iFoVjlYEi2nh4XZzgEe9_u0wYo1EmX8/view?usp=sharing](url)

I cheated on the particle systems but everything else is affected directly by the MIDI information.

I would really like to use your MIDI add-on to create abstract animations for orchestral pieces but that would require a lot of work, hence the simple experiments first.

Have you (or anyone else you know) used this to create 3D content for big ensembles?

**Comment by Cornerback24** [comment]
Hey, I haven't created any big ensemble animation with this add-on (since I've probably spent more time on developing the add-on than creating animations in Blender).  I think the closest to big ensemble content created with this add-on that I know of might be this https://www.youtube.com/watch?v=3rxoNzjU1cw

**Comment by ElixirHunter** [comment]
I like it - well done to them for that 👍

It is more in the animusic tradition of using 'instrument looking' animations. I am thinking of a more abstract set up and yet controlled by MIDI.

I have a some projects that I need to turn to now so will be going 'dark' for a while but hopefully I will return to this again one day.

Thank you for all your responses and **_Good Luck_** with all your projects Cornerback 😁

**Reply by ElixirHunter** [reply]
Hi CornerBack,

Hope you are well.

I am going to try and have another dabble with your plug in.

One thing that would be really good for me is to be able to choose other MIDI 'continuous controllers'  i.e. CC DATA.

For example I have instruments that get louder and quieter but I do not use the velocity DATA (how hard you hit the key) but instead other controllers (such as expression, aftertouch etc.)

It would be great if instead of just having a choice of pitch, velocity and length you could choose any CC DATA as a control. Do you think this could be possible?

Best wishes,
Elixir

**Comment by Cornerback24** [comment]
Hi, I think that could be a challenge because right now the plugin does not have any continuous data, only note specific data. However I think it would be possible.

What would you want it to do, would it add keyframes in the graph editor whenever the CC data changes?

**Comment by ElixirHunter** [comment]
Hi, thanks for the quick repsonse.

I will explain what I am thinking when I can get to my
computer.......probably sometime on Sunday.

Cheers for now.
